resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/AliasManager.java

```diff
@@ -624,27 +621,40 @@ public void updateAliases(IResource resource, IFileStore location, int depth, IP
 	/**
 	 * Process any structural changes that have occurred since the last alias
 	 * request.
+	 *
+	 * @return <code>true</code> if some structural changes where processed,
+	 *         <code>false</code> if no structural changes were detected
 	 */
-	private void updateStructureChanges() {
+	/*
+	 * This method is synchronized as it calls
+	 * buildLocationsMap/addToLocationsMap/addToLocationsMap which are not meant to
+	 * run in parallel. Incoming events/notifications that cause a structural change
+	 * can come from different threads, but we only want a single thread to process
+	 * them at a time.
+	 */
+	private synchronized boolean checkStructuralChanges() {
 		boolean hadChanges = false;
-		if (changedProjects) {
+		while (changedProjects) {
```

Comment:
previously it was either buildLocationsMap or addToLocationsMap now both can happen and add the same project - which is also a resource? That may increase nonDefaultResourceCount one time too much?

Reply:
I need to investigate. Do you have a clearer idea of how things can be wrong? I hope we can make a test case for it.

Reply:
already the current implementation looks buggy or tricky as hell: 
only handleEvent(LifecycleEvent)  does nonDefaultResourceCount--
but both  handleEvent(LifecycleEvent) and resourceChanged(IResourceChangeEvent event)
could lead to nonDefaultResourceCount++

i guess the Trick was that buildLocationsMap() resets it to 0 - which does not hold anymore in your patch

Reply:
You're right, the while and removing the else make it more fragile while mot much improving/fixing anything.
I've restored to something that's closer from initial state and added comments to keep that in mind for further work.